README.md
pypinfo: View PyPI download statistics with ease.
=================================================

[![version](https://img.shields.io/pypi/v/pypinfo.svg?style=flat-square)](https://pypi.org/project/pypinfo)
[![python](https://img.shields.io/pypi/pyversions/pypinfo.svg?style=flat-square)](https://pypi.org/project/pypinfo)
[![license](https://img.shields.io/badge/license-MIT-blue.svg?style=flat-square)](https://en.wikipedia.org/wiki/MIT_License)
[![license](https://img.shields.io/badge/code%20style-black-000000.svg?style=flat-square)](https://github.com/psf/black)

pypinfo is a simple CLI to access `PyPI`_ download statistics via Google's BigQuery.

* `Usage <#usage>`_
* `Installation <#installation>`_
* `Credits <#credits>`_

<details>
  <summary>Usage</summary>
  
    Usage
    -----

    ```console
    $ pypinfo
        Usage: pypinfo [OPTIONS] [PROJECT] [FIELDS]... COMMAND [ARGS]...

          Valid fields are:

          project | version | file | pyversion | percent3 | percent2 | impl | impl-version |

          openssl | date | month | year | country | installer | installer-version |

          setuptools-version | system | system-release | distro | distro-version | cpu |

          libc | libc-version

        Options:
          -a, --auth TEXT         Path to Google credentials JSON file.
          --run / --test          --test simply prints the query.
          -j, --json              Print data as JSON, with keys `rows` and `query`.
          -i, --indent INTEGER    JSON indentation level.
          -t, --timeout INTEGER   Milliseconds. Default: 120000 (2 minutes)
          -l, --limit TEXT        Maximum number of query results. Default: 10
          -d, --days TEXT         Number of days in the past to include. Default: 30
          -sd, --start-date TEXT  Must be negative or YYYY-MM[-DD]. Default: -31
          -ed, --end-date TEXT    Must be negative or YYYY-MM[-DD]. Default: -1
          -m, --month TEXT        Shortcut for -sd & -ed for a single YYYY-MM month.
          -w, --where TEXT        WHERE conditional. Default: file.project = "project"
          -o, --order TEXT        Field to order by. Default: download_count
          --all                   Show downloads by all installers, not only pip.
          -pc, --percent          Print percentages.
          -md, --markdown         Output as Markdown.
          -v, --verbose           Print debug messages to stderr.
          --version               Show the version and exit.
          -h, --help              Show this message and exit.
    ```

    pypinfo accepts 0 or more options, followed by exactly 1 project, followed by
    0 or more fields. By default only the last 30 days are queried. Let's take a
    look at some examples!

    Tip: If queries are resulting in NoneType errors, increase timeout.

    Downloads for a project
    ^^^^^^^^^^^^^^^^^^^^^^^

    ```console

        $ pypinfo requests
        Served from cache: False
        Data processed: 2.83 GiB
        Data billed: 2.83 GiB
        Estimated cost: $0.02

        | download_count |
        | -------------- |
        |    116,353,535 |
    ```

    All downloads
    ^^^^^^^^^^^^^

    ```console

        $ pypinfo ""
        Served from cache: False
        Data processed: 116.15 GiB
        Data billed: 116.15 GiB
        Estimated cost: $0.57

        | download_count |
        | -------------- |
        |  8,642,447,168 |
    ```

    Downloads for a project by Python version
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

    ```console

        $ pypinfo django pyversion
        Served from cache: False
        Data processed: 967.33 MiB
        Data billed: 968.00 MiB
        Estimated cost: $0.01

        | python_version | download_count |
        | -------------- | -------------- |
        | 3.8            |      1,735,967 |
        | 3.6            |      1,654,871 |
        | 3.7            |      1,326,423 |
        | 2.7            |        876,621 |
        | 3.9            |        524,570 |
        | 3.5            |        258,609 |
        | 3.4            |         12,769 |
        | 3.10           |          3,050 |
        | 3.3            |            225 |
        | 2.6            |            158 |
        | Total          |      6,393,263 |
    ```

    All downloads by country code
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

    ```console

        $ pypinfo "" country
        Served from cache: False
        Data processed: 150.40 GiB
        Data billed: 150.40 GiB
        Estimated cost: $0.74

        | country | download_count |
        | ------- | -------------- |
        | US      |  6,614,473,568 |
        | IE      |    336,037,059 |
        | IN      |    192,914,402 |
        | DE      |    186,968,946 |
        | NL      |    182,691,755 |
        | None    |    141,753,357 |
        | BE      |    111,234,463 |
        | GB      |    109,539,219 |
        | SG      |    106,375,274 |
        | FR      |     86,036,896 |
        | Total   |  8,068,024,939 |
    ```

    Downloads for a project by system and distribution
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

    ```console

        $ pypinfo cryptography system distro
        Served from cache: False
        Data processed: 2.52 GiB
        Data billed: 2.52 GiB
        Estimated cost: $0.02

        | system_name | distro_name                     | download_count |
        | ----------- | ------------------------------- | -------------- |
        | Linux       | Ubuntu                          |     19,524,538 |
        | Linux       | Debian GNU/Linux                |     11,662,104 |
        | Linux       | Alpine Linux                    |      3,105,553 |
        | Linux       | Amazon Linux AMI                |      2,427,975 |
        | Linux       | Amazon Linux                    |      2,374,869 |
        | Linux       | CentOS Linux                    |      1,955,181 |
        | Windows     | None                            |      1,522,069 |
        | Linux       | CentOS                          |        568,370 |
        | Darwin      | macOS                           |        489,859 |
        | Linux       | Red Hat Enterprise Linux Server |        296,858 |
        | Total       |                                 |     43,927,376 |
    ```

    Most popular projects in the past year
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

    ```console

        $ pypinfo --days 365 "" project
        Served from cache: False
        Data processed: 1.69 TiB
        Data billed: 1.69 TiB
        Estimated cost: $8.45

        | project         | download_count |
        | --------------- | -------------- |
        | urllib3         |  1,382,528,406 |
        | six             |  1,172,798,441 |
        | botocore        |  1,053,169,690 |
        | requests        |    995,387,353 |
        | setuptools      |    992,794,567 |
        | certifi         |    948,518,394 |
        | python-dateutil |    934,709,454 |
        | idna            |    929,781,443 |
        | s3transfer      |    877,565,186 |
        | chardet         |    854,744,674 |
        | Total           | 10,141,997,608 |
    ```

    Downloads between two YYYY-MM-DD dates
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

    ```console

        $ pypinfo --start-date 2018-04-01 --end-date 2018-04-30 setuptools
        Served from cache: False
        Data processed: 571.37 MiB
        Data billed: 572.00 MiB
        Estimated cost: $0.01

        | download_count |
        | -------------- |
        |      8,972,826 |
    ```

    Downloads between two YYYY-MM dates
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

    - A yyyy-mm ``--start-date`` defaults to the first day of the month
    - A yyyy-mm ``--end-date`` defaults to the last day of the month

    ```console

        $ pypinfo --start-date 2018-04 --end-date 2018-04 setuptools
        Served from cache: False
        Data processed: 571.37 MiB
        Data billed: 572.00 MiB
        Estimated cost: $0.01

        | download_count |
        | -------------- |
        |      8,972,826 |
    ```

    Downloads for a single YYYY-MM month
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

    ```console

        $ pypinfo --month 2018-04 setuptools
        Served from cache: False
        Data processed: 571.37 MiB
        Data billed: 572.00 MiB
        Estimated cost: $0.01

        | download_count |
        | -------------- |
        |      8,972,826 |
    ```

    Percentage of Python 3 downloads of the top 100 projects in the past year
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

    Let's use ``--test`` to only see the query instead of sending it.

    ```console

        $ pypinfo --test --days 365 --limit 100 "" project percent3
        SELECT
          file.project as project,
          ROUND(100 * SUM(CASE WHEN REGEXP_EXTRACT(details.python, r"^([^\.]+)") = "3" THEN 1 ELSE 0 END) / COUNT(*), 1) as percent_3,
          COUNT(*) as download_count,
        FROM `bigquery-public-data.pypi.file_downloads`
        WHERE timestamp BETWEEN TIMESTAMP_ADD(CURRENT_TIMESTAMP(), INTERVAL -366 DAY) AND TIMESTAMP_ADD(CURRENT_TIMESTAMP(), INTERVAL -1 DAY)
          AND details.installer.name = "pip"
        GROUP BY
          project
        ORDER BY
          download_count DESC
        LIMIT 100
    ```

    Downloads for a given version
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

    pypinfo supports `PEP 440 version matching <https://peps.python.org/pep-0440/#version-matching>`_.

    We can use it to query stats on a given major version.

    ```console

        $ pypinfo -pc 'pip==21.*' pyversion version
        Served from cache: False
        Data processed: 34.45 MiB
        Data billed: 35.00 MiB
        Estimated cost: $0.01

        | python_version | version | percent | download_count |
        | -------------- | ------- | ------- | -------------- |
        | 3.6            | 21.3.1  |  78.74% |         10,430 |
        | 3.8            | 21.3.1  |   7.81% |          1,034 |
        | 3.7            | 21.2.1  |   3.59% |            476 |
        | 3.7            | 21.3.1  |   2.60% |            345 |
        | 3.7            | 21.0.1  |   2.25% |            298 |
        | 3.8            | 21.0.1  |   1.58% |            209 |
        | 3.8            | 21.2.1  |   1.42% |            188 |
        | 3.7            | 21.1.2  |   0.81% |            107 |
        | 3.9            | 21.3.1  |   0.69% |             92 |
        | 3.8            | 21.1.1  |   0.51% |             67 |
        | Total          |         |         |         13,246 |
    ```

    We can also use it to query stats on an exact version:

    ```console

        $ pypinfo -pc 'numpy==1.23rc3' pyversion version
        Served from cache: False
        Data processed: 34.01 MiB
        Data billed: 35.00 MiB
        Estimated cost: $0.01

        | python_version | version   | percent | download_count |
        | -------------- | --------- | ------- | -------------- |
        | 3.9            | 1.23.0rc3 |  63.33% |             38 |
        | 3.8            | 1.23.0rc3 |  28.33% |             17 |
        | 3.10           | 1.23.0rc3 |   8.33% |              5 |
        | Total          |           |         |             60 |
    ```

</details>


Installation
------------

pypinfo is distributed on `PyPI`_ as a universal wheel and is available on
Linux/macOS and Windows and supports Python 3.7+.

This is relatively painless, I swear.

**Create project**

1. Go to `<https://bigquery.cloud.google.com>`_.
2. Sign up if you haven't already. The first TB of queried data each month
   is free. Each additional TB is $5.

3. Go to `<https://console.developers.google.com/cloud-resource-manager>`_ and click
   CREATE PROJECT if you don't already have one:
    
   ![create](https://user-images.githubusercontent.com/1324225/47172949-6f4ea880-d315-11e8-8587-8b8117efeae9.png "CREATE PROJECT")

4. This takes you to `<https://console.developers.google.com/projectcreate>`_. Fill out
   the form and click CREATE. Any name is fine, but I recommend you choose something to
   do with PyPI like pypinfo. This way you know what the project is designated for:

   ![click](https://user-images.githubusercontent.com/1324225/47173020-986f3900-d315-11e8-90ab-4b2ecd85b88e.png) 

5. The next page should show your new project. If not, reload the page and select from the top menu:

   ![show](https://user-images.githubusercontent.com/1324225/47173170-0b78af80-d316-11e8-879e-01f34e139b80.png)
   

**Enable BigQuery API**

6. Go to `<https://console.cloud.google.com/apis/api/bigquery-json.googleapis.com/overview>`_
   and make sure the correct project is chosen using the drop-down on top. Click
   the ENABLE button:

   ![enable](https://user-images.githubusercontent.com/1324225/47173408-a6718980-d316-11e8-94c2-a17ff54fc389.png)

7. After enabling, click CREATE CREDENTIALS:

   ![credentials](https://user-images.githubusercontent.com/1324225/47173432-bc7f4a00-d316-11e8-8152-6a0e6cfab70f.png)

8. Choose the "BigQuery API" and "No, I'm not using them":

   ![bigquery-api](https://user-images.githubusercontent.com/1324225/47173510-ec2e5200-d316-11e8-8508-2bfbb8f6b02f.png)

9. Fill in a name, and select role "BigQuery User" (if the "BigQuery" is not an option
   in the list, wait 15-20 minutes and try creating the credentials again), and select a JSON key:

   ![bigquery-user](https://user-images.githubusercontent.com/1324225/47173576-18e26980-d317-11e8-8bfe-e4775d965e32.png)

10. Click continue and the JSON will download to your computer. Note the download location. Move the file wherever you want:

   ![download](https://user-images.githubusercontent.com/1324225/47173614-331c4780-d317-11e8-9ed2-fc76557a2bf6.png)

11. ``pip install pypinfo``
12. ``pypinfo --auth path/to/your_credentials.json``, or set an environment variable
    ``GOOGLE_APPLICATION_CREDENTIALS`` that points to the file.

Credits
-------

- `Donald Stufft <https://github.com/dstufft>`_ for maintaining `PyPI`_ all
  these years.
- `Google <https://github.com/google>`_ for donating BigQuery capacity to
  `PyPI`_.
- `Paul Kehrer <https://github.com/reaperhulk>`_ for his
  `awesome blog post <https://langui.sh/2016/12/09/data-driven-decisions>`_.

.. _PyPI: https://pypi.org


CHANGELOG.rst
Changelog
---------

Important changes are emphasized.

Unreleased
^^^^^^^^^^

20.0.0
^^^^^^

- Add support for Python 3.10
- Add ``-h`` as help option
- Treat and validate numeric CLI arguments as numbers
- Replace appdirs with platformdirs
- Fix ResourceWarnings

19.0.0
^^^^^^

- Update dataset to the new Google-hosted location

18.0.1
^^^^^^

- Fix usage of date ranges

18.0.0
^^^^^^

- Use the clustered data table and standard SQL for lower query costs

17.0.0
^^^^^^

- Add support for libc & libc-version fields

16.0.2
^^^^^^

- Update TinyDB and Tinyrecord dependencies for compatibility

16.0.1
^^^^^^

- Pin TinyDB<4, Tinyrecord does not yet support TinyDB v4

16.0.0
^^^^^^

- Allow yyyy-mm[-dd] ``--start-date`` and ``--end-date``:

  - A yyyy-mm ``--start-date`` defaults to the first day of the month
  - A yyyy-mm ``--end-date`` defaults to the last day of the month

- Add ``--month`` as a shortcut to ``--start-date`` and ``--end-date``
  for a single yyyy-mm month

- Add ``--verbose`` option to print credentials location

- Update installation instructions

- Enforce ``black`` code style

15.0.0
^^^^^^

- Allow yyyy-mm-dd dates
- Add ``--all`` option, default to only showing downloads via pip
- Add download total row

14.0.0
^^^^^^

- Added new ``file`` field!

13.0.0
^^^^^^

- Added ``last_update`` JSON key, which is a UTC timestamp.

12.0.0
^^^^^^

- **Breaking:** JSON output is now a mapping with keys ``rows``, which is all the
  data that was previously outputted, and ``query``, which is relevant metadata.
- Increased the resolution of percentages.

11.0.0
^^^^^^

- Fixed JSON output.

10.0.0
^^^^^^

- Fixed custom field ordering.

9.0.0
^^^^^

- Added new BigQuery usage stats.
- Lowered the default number of results to ``10`` from ``20``.
- Updated examples.
- Fixed table formatting regression.

8.0.0
^^^^^

- Updated ``google-cloud-bigquery`` dependency.

7.0.0
^^^^^

- Output table is now in Markdown format for easy copying to GitHub issues and PRs.

6.0.0
^^^^^

- Updated ``google-cloud-bigquery`` dependency.

5.0.0
^^^^^

- Numeric output (non-json) is now prettier (thanks `hugovk <https://github.com/hugovk>`_)
- You can now filter results for only pip installs with the ``--pip`` flag
  (thanks `hugovk <https://github.com/hugovk>`_)

4.0.0
^^^^^

- ``--order`` now works with all fields (thanks `Brian Skinn <https://github.com/bskinn>`_)
- Updated installation docs (thanks `Brian Skinn <https://github.com/bskinn>`_)

3.0.1
^^^^^

- Fix: project names are now normalized to adhere to
  `PEP 503 <https://www.python.org/dev/peps/pep-0503>`_.

3.0.0
^^^^^

- **Breaking:** ``--json`` option is now just a flag and prints output as prettified JSON.

2.0.0
^^^^^

- Added ``--json`` path option.

1.0.0
^^^^^

- Initial release

.. _PyPI: https://pypi.org


tests/test_db.py
from pypinfo import db

CREDS_FILE = '/path/to/creds_file.json'


def test_get_credentials(tmp_path):
    # Arrange
    db.DB_FILE = str(tmp_path / 'db.json')  # Mock

    # Assert
    assert db.get_credentials() is None


def test_set_credentials(tmp_path):
    # Arrange
    db.DB_FILE = str(tmp_path / 'db.json')  # Mock

    # Act
    db.set_credentials(CREDS_FILE)


def test_set_credentials_twice(tmp_path):
    # Arrange
    db.DB_FILE = str(tmp_path / 'db.json')  # Mock

    # Act
    db.set_credentials(CREDS_FILE)
    db.set_credentials(CREDS_FILE)


def test_round_trip(tmp_path):
    # Arrange
    db.DB_FILE = str(tmp_path / 'db.json')  # Mock

    # Act
    db.set_credentials(CREDS_FILE)

    # Assert
    assert db.get_credentials() == CREDS_FILE


def test_get_credentials_table(tmp_path):
    db.DB_FILE = str(tmp_path / 'db.json')
    with db.get_credentials_table() as table:
        assert not table._storage._handle.closed
        with db.get_credentials_table(table) as table2:
            assert table2 is table
        assert not table._storage._handle.closed
    assert table._storage._handle.closed


pypinfo/db.py
import contextlib
import os
from typing import Iterator, Optional

from platformdirs import user_data_dir
from tinydb import TinyDB, Query, where
from tinydb.table import Table
from tinyrecord import transaction

DB_FILE = os.path.join(user_data_dir('pypinfo', appauthor=False), 'db.json')


@contextlib.contextmanager
def get_credentials_table(table: Optional[Table] = None) -> Iterator[Table]:
    if table is None:
        with TinyDB(DB_FILE, create_dirs=True) as db:
            yield db.table('credentials')
    else:
        yield table


def get_credentials(table: Optional[Table] = None) -> Optional[str]:
    with get_credentials_table(table) as table:
        query = table.search(Query().path.exists())
        return query[0]['path'] if query else None


def set_credentials(creds_file: str) -> None:
    with get_credentials_table() as table:
        exists = get_credentials(table)
        with transaction(table) as tr:
            if exists:
                tr.update({'path': creds_file}, where('path').exists())
            else:
                tr.insert({'path': creds_file})


pypinfo/fields.py
from collections import namedtuple

Field = namedtuple('Field', ('name', 'data'))
Downloads = Field('download_count', 'COUNT(*)')
Date = Field('download_date', 'FORMAT_TIMESTAMP("%Y-%m-%d", timestamp)')
Month = Field('download_month', 'FORMAT_TIMESTAMP("%Y-%m", timestamp)')
Year = Field('download_year', 'FORMAT_TIMESTAMP("%Y", timestamp)')
Country = Field('country', 'country_code')
Project = Field('project', 'file.project')
Version = Field('version', 'file.version')
File = Field('file', 'file.filename')
PythonVersion = Field('python_version', r'REGEXP_EXTRACT(details.python, r"^([^\.]+\.[^\.]+)")')
Percent3 = Field(
    'percent_3',
    r'ROUND(100 * SUM(CASE WHEN REGEXP_EXTRACT(details.python, r"^([^\.]+)") = "3" THEN 1 ELSE 0 END) / COUNT(*), 1)',
)
Percent2 = Field(
    'percent_2',
    r'ROUND(100 * SUM(CASE WHEN REGEXP_EXTRACT(details.python, r"^([^\.]+)") = "2" THEN 1 ELSE 0 END) / COUNT(*), 1)',
)
Implementation = Field('implementation', 'details.implementation.name')
ImplementationVersion = Field('impl_version', r'REGEXP_EXTRACT(details.implementation.version, r"^([^\.]+\.[^\.]+)")')
OpenSSLVersion = Field('openssl_version', 'REGEXP_EXTRACT(details.openssl_version, r"^OpenSSL ([^ ]+) ")')
Installer = Field('installer_name', 'details.installer.name')
InstallerVersion = Field('installer_version', 'details.installer.version')
SetuptoolsVersion = Field('setuptools_version', 'details.setuptools_version')
System = Field('system_name', 'details.system.name')
SystemRelease = Field('system_release', 'details.system.release')
Distro = Field('distro_name', 'details.distro.name')
DistroVersion = Field('distro_version', 'details.distro.version')
CPU = Field('cpu', 'details.cpu')
Libc = Field('libc_name', 'details.distro.libc.lib')
LibcVersion = Field('libc_version', 'details.distro.libc.version')

AGGREGATES = {Percent3, Percent2}


pypinfo/cli.py
from decimal import ROUND_UP, Decimal
from typing import List

import click
from binary import TEBIBYTE, convert_units

from pypinfo.core import (
    add_percentages,
    add_download_total,
    build_query,
    create_client,
    create_config,
    format_json,
    month_ends,
    parse_query_result,
    tabulate,
)
from pypinfo.db import get_credentials, set_credentials
from pypinfo.fields import (
    Project,
    Date,
    Month,
    Year,
    Country,
    Version,
    File,
    PythonVersion,
    Percent3,
    Percent2,
    Installer,
    InstallerVersion,
    SetuptoolsVersion,
    System,
    SystemRelease,
    Implementation,
    ImplementationVersion,
    OpenSSLVersion,
    Distro,
    DistroVersion,
    CPU,
    Libc,
    LibcVersion,
)

CONTEXT_SETTINGS = {
    'help_option_names': ('-h', '--help'),
    'max_content_width': 300,
}
FIELD_MAP = {
    'project': Project,
    'version': Version,
    'file': File,
    'pyversion': PythonVersion,
    'percent3': Percent3,
    'percent2': Percent2,
    'impl': Implementation,
    'impl-version': ImplementationVersion,
    'openssl': OpenSSLVersion,
    'date': Date,
    'month': Month,
    'year': Year,
    'country': Country,
    'installer': Installer,
    'installer-version': InstallerVersion,
    'setuptools-version': SetuptoolsVersion,
    'system': System,
    'system-release': SystemRelease,
    'distro': Distro,
    'distro-version': DistroVersion,
    'cpu': CPU,
    'libc': Libc,
    'libc-version': LibcVersion,
}
TIER_COST = 5
TB = Decimal(TEBIBYTE)
TO_CENTS = Decimal('0.00')


@click.group(invoke_without_command=True, context_settings=CONTEXT_SETTINGS)
@click.argument('project', required=False)
@click.argument('fields', nargs=-1, required=False)
@click.option('--auth', '-a', help='Path to Google credentials JSON file.')
@click.option('--run/--test', default=True, help='--test simply prints the query.')
@click.option('--json', '-j', is_flag=True, help='Print data as JSON, with keys `rows` and `query`.')
@click.option('--indent', '-i', type=int, help='JSON indentation level.')
@click.option('--timeout', '-t', type=int, default=120000, help='Milliseconds. Default: 120000 (2 minutes)')
@click.option('--limit', '-l', type=int, help='Maximum number of query results. Default: 10')
@click.option('--days', '-d', type=int, help='Number of days in the past to include. Default: 30')
@click.option('--start-date', '-sd', help='Must be negative or YYYY-MM[-DD]. Default: -31')
@click.option('--end-date', '-ed', help='Must be negative or YYYY-MM[-DD]. Default: -1')
@click.option('--month', '-m', help='Shortcut for -sd & -ed for a single YYYY-MM month.')
@click.option('--where', '-w', help='WHERE conditional. Default: file.project = "project"')
@click.option('--order', '-o', help='Field to order by. Default: download_count')
@click.option('--all', 'all_installers', is_flag=True, help='Show downloads by all installers, not only pip.')
@click.option('--percent', '-pc', is_flag=True, help='Print percentages.')
@click.option('--markdown', '-md', is_flag=True, help='Output as Markdown.')
@click.option('--verbose', '-v', is_flag=True, help='Print debug messages to stderr.')
@click.version_option()
@click.pass_context
def pypinfo(
    ctx: click.Context,
    project: str,
    fields: List[str],
    auth: str,
    run: bool,
    json: bool,
    indent: int,
    timeout: int,
    limit: int,
    days: int,
    start_date: str,
    end_date: str,
    month: str,
    where: str,
    order: str,
    all_installers: bool,
    percent: bool,
    markdown: bool,
    verbose: bool,
) -> None:
    """Valid fields are:\n
    project | version | file | pyversion | percent3 | percent2 | impl | impl-version |\n
    openssl | date | month | year | country | installer | installer-version |\n
    setuptools-version | system | system-release | distro | distro-version | cpu |\n
    libc | libc-version
    """
    if auth:
        set_credentials(auth)
        click.echo(f'Credentials location set to "{get_credentials()}".')
        return

    if verbose:
        click.echo(f'Credentials location set to "{get_credentials()}".', err=True)

    if project is None and not fields:
        click.echo(ctx.get_help())
        return

    parsed_fields = []
    for field in fields:
        parsed = FIELD_MAP.get(field)
        if parsed is None:
            raise ValueError(f'"{field}" is an unsupported field.')
        parsed_fields.append(parsed)

    order_name = order
    parsed_order = FIELD_MAP.get(order)
    if parsed_order:
        order_name = parsed_order.name
        parsed_fields.insert(0, parsed_order)

    if month:
        start_date, end_date = month_ends(month)

    built_query = build_query(
        project,
        parsed_fields,
        limit=limit,
        days=days,
        start_date=start_date,
        end_date=end_date,
        where=where,
        order=order_name,
        pip=not all_installers,
    )

    if run:
        with create_client(get_credentials()) as client:
            query_job = client.query(built_query, job_config=create_config())
            query_rows = query_job.result(timeout=timeout // 1000)
            rows = parse_query_result(query_job, query_rows)

        # Cached
        from_cache = not not query_job.cache_hit

        # Processed
        bytes_processed = query_job.total_bytes_processed or 0
        processed_amount, processed_unit = convert_units(bytes_processed)

        # Billed
        bytes_billed = query_job.total_bytes_billed or 0
        billed_amount, billed_unit = convert_units(bytes_billed)

        # Cost
        billing_tier = query_job.billing_tier or 1
        estimated_cost = Decimal(TIER_COST * billing_tier) / TB * Decimal(bytes_billed)
        estimated_cost_str = str(estimated_cost.quantize(TO_CENTS, rounding=ROUND_UP))

        if len(rows) == 1 and not json:
            # Only headers returned
            click.echo("No data returned, check project name")
            return

        if percent:
            rows = add_percentages(rows, include_sign=not json)

        # Only for tables, and if more than the header row + a single data row
        if len(rows) > 2 and not json:
            rows = add_download_total(rows)

        if not json:
            click.echo(f'Served from cache: {from_cache}')
            click.echo(f'Data processed: {processed_amount:.2f} {processed_unit}')
            click.echo(f'Data billed: {billed_amount:.2f} {billed_unit}')
            click.echo(f'Estimated cost: ${estimated_cost_str}')

            click.echo()
            click.echo(tabulate(rows, markdown))
        else:
            query_info = {
                'cached': from_cache,
                'bytes_processed': bytes_processed,
                'bytes_billed': bytes_billed,
                'estimated_cost': estimated_cost_str,
            }
            click.echo(format_json(rows, query_info, indent))
    else:
        click.echo(built_query)
